File: interviews/leetcode/leetcode-75/binary-tree-dfs/872. Leaf-Similar Trees/solution_iterative.py
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        # Helper function to collect the leaf nodes in a tree iteratively

        # Get the leaf nodes for both trees
        leaves1 = getLeaves(root1)
        leaves2 = getLeaves(root2)

        # Compare the leaf nodes
        return leaves1 == leaves2

# Helper function to collect the leaf nodes in a tree iteratively
def getLeaves(root):
    leaves = []
    if not root:
        return leaves
    stack = [root]
    while stack:
        node = stack.pop()
        if not node.left and not node.right:
            leaves.append(node.val)
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)
    return leaves

# Helper function to create a binary tree from a list (for testing purposes)
def create_binary_tree(lst):
    if not lst:
        return None
    root = TreeNode(lst[0])
    queue = [root]
    i = 1
    while i < len(lst):
        node = queue.pop(0)
        if lst[i] is not None:
            node.left = TreeNode(lst[i])
            queue.append(node.left)
        i += 1
        if i < len(lst) and lst[i] is not None:
            node.right = TreeNode(lst[i])
            queue.append(node.right)
        i += 1
    return root

# Test the solution
# if __name__ == "__main__":
#     solution := Solution{}
#     root1 := create_binary_tree([]int{3, 5, 1, 6, 2, 9, 8, nil, nil, 7, 4})
#     root2 := create_binary_tree([]int{3, 5, 1, 6, 7, 4, 2, nil, nil, nil, nil, 9, 8})
#     result := solution.leafSimilar(root1, root2)
#     fmt.Println(result)  // Output: true

File: interviews/leetcode/leetcode-75/binary-tree-dfs/872. Leaf-Similar Trees/solution_iterative.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Helper function to collect the leaf nodes of a binary tree iteratively
func getLeaves(root *TreeNode) []int {
	var leaves []int
	if root == nil {
		return leaves
	}

	stack := []*TreeNode{root}
	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		// If it's a leaf, add to the leaves list
		if node.Left == nil && node.Right == nil {
			leaves = append(leaves, node.Val)
		}

		// Push right and left children to the stack
		if node.Right != nil {
			stack = append(stack, node.Right)
		}
		if node.Left != nil {
			stack = append(stack, node.Left)
		}
	}

	return leaves
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	// Get the leaf nodes for both trees
	leaves1 := getLeaves(root1)
	leaves2 := getLeaves(root2)

	// Compare the leaf nodes of both trees
	if len(leaves1) != len(leaves2) {
		return false
	}
	for i := 0; i < len(leaves1); i++ {
		if leaves1[i] != leaves2[i] {
			return false
		}
	}
	return true
}

// Helper function to create a binary tree from a slice (for testing purposes)
func createBinaryTree(lst []interface{}) *TreeNode {
	if len(lst) == 0 || lst[0] == nil {
		return nil
	}
	root := &TreeNode{Val: lst[0].(int)}
	queue := []*TreeNode{root}
	i := 1
	for i < len(lst) {
		node := queue[0]
		queue = queue[1:]
		if lst[i] != nil {
			node.Left = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(lst) && lst[i] != nil {
			node.Right = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

/*
func main() {
	// Test the solution
	root1 := createBinaryTree([]interface{}{3, 5, 1, 6, 2, 9, 8, nil, nil, 7, 4})
	root2 := createBinaryTree([]interface{}{3, 5, 1, 6, 7, 4, 2, nil, nil, nil, nil, 9, 8})

	result := leafSimilar(root1, root2)
	fmt.Println(result) // Output: true
}
*/
